File: problem78.py
import math
ONE_MIL = 1000000

partitions = [0]*ONE_MIL
partitions[0] = 1

def get_partitions(k):
    if k < 0:
        return 0
    if k >= ONE_MIL:
        return -1
    return partitions[k]

def solve():
    n = 1
    while n < ONE_MIL:
        total = 0
        lb = int(math.ceil(-(math.sqrt(24*n+1)+1.0)/6))
        ub = int(math.floor((math.sqrt(24*n+1)-1.0)/6))
        for k in range(lb, ub + 1):
            if k == 0:
                continue
            coeff = 1
            if k % 2 == 0:
                coeff = -1
            total += coeff*get_partitions(n - k*(3*k+1)//2) % ONE_MIL
            total %= ONE_MIL
        if total == 0:
            return n
        partitions[n] = total
        n += 1

File: problem78.c
#include<stdio.h>
#include<math.h>

#define ONE_MIL 1000000

int partitions[ONE_MIL];

inline int get_partitions(int k) {
    if (k < 0) {
        return 0;
    }
    if (k >= ONE_MIL) {
        return -1;
    }
    return partitions[k];
}

inline int pos_mod(int i, int n) {
    return (i % n + n) % n;
}

int main() {
    partitions[0] = 1;
    int n = 1;
    int total, coeff, k, lb, ub;
    while (n < ONE_MIL) {
        //if (n%1000 == 0) printf("%d\n", n);
        total = 0;
        lb = (int)ceil(-(sqrt(24*n+1)+1.0)/6);
        ub = (int)floor((sqrt(24*n+1)-1.0)/6);
        //printf("%d <= k <= %d; ", lb, ub);
        for (k = lb; k <= ub; k++) {
            if (k == 0) continue;
            if (k % 2 == 0) coeff = -1;
            else coeff = 1;
            total += pos_mod(coeff*get_partitions(n - k*(3*k+1)/2), ONE_MIL);
            total = pos_mod(total, ONE_MIL);
        }
        //printf("p(%d) = %d\n", n, total);
        //total = total;
        if (total == 0) {
            printf("SOLUTION: %d\n", n);
            return 0;
        }
        partitions[n] = total;
        n++;
    }
}